domain/charm/state/state.go

```diff
@@ -1097,12 +1120,86 @@ ORDER BY array_index ASC;
 	return result, nil
 }
 
+func (s *State) checkSetCharmExists(ctx context.Context, tx *sqlair.TX, id corecharm.ID, name string, revision int) error {
+	selectQuery := `
+SELECT charm.uuid AS &charmID.*
+FROM charm
+LEFT JOIN charm_origin ON charm.uuid = charm_origin.charm_uuid
+WHERE charm.name = $charmNameRevision.name AND charm_origin.revision = $charmNameRevision.revision
+
+	`
+	selectStmt, err := s.Prepare(selectQuery, charmID{}, charmNameRevision{})
```

Comment:
Can create the structs before the prepare and the same one to prepare and query.